DataDog/dd-trace-rb
# Add Remote Configuration boot tags

**2.0 Upgrade Guide notes**


**What does this PR do?**

- Add tags to record what happened WRT RC boot on a specific request
- Add tags to record RC state for a specific request
- Improve Barrier to return the cause of wait_once being unblocked
- Add tuneable barrier timeout via configuration

**Motivation:**

RC boot waiting being best-effort, it becomes important to be able to diagnose specific requests behaving in unexpected ways WRT remote configuration settings, e.g some select requests not being blocked by a denylist entry.

By adding tags recording what happened with remote configuration boot on a specific request we can diagnose why a specific request behaved a certain way instead of relying on guesswork.

This is fundamentally different from gathering statistical metrics pushed via telemetry (another planned thing).

**Additional Notes:**

Since these only concern specific information for boot, these tags are present only for requests that are subject to waiting during the boot process.

Recording the time as a metric for all boot requests also allows us to build knowledge of what a sane default value would be.

~A PR to allow configuration of the barrier timeout is coming up next.~ Timeout is configurable.

**How to test the change?**

This starts as a draft to gather feedback, ~unit tests pending~ Unit tests are implemented.

Manual check:

- either use some tool to delay RC response beyond timeout or configure to be absurdly low such as 0.01, observe timeout tag value being set
- with default value (1.0), observe ready tag value being set
- observe that time is recorded in all waiting cases
- observe tags are not set for post-boot requests
- observe that tags are not set when RC is disabled



**For Datadog employees:**
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

Unsure? Have a question? Request a review!
